smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,355 @@
+            {
+                agent_id: gym.spaces.Box(
+                    np.array([0, 0, -1]), np.array([+1, +1, +1]), dtype=np.float32
+                )  # throttle, break, steering
+                for agent_id in ["Agent1", "Agent2"]
+            }
+        )
+        observation_space = gym.spaces.Box(
+            low=-1, high=1, shape=(256, 256, 3), dtype=np.float32
+        )
+
+        super(AsyncVectorEnv, self).__init__(
+            num_envs=len(env_constructors),
+            observation_space=observation_space,
+            action_space=action_space,
+        )
+
+        # Seed all the environment
+        self.seed(seed)
+
+    def seed(self, seed: int) -> Tuple[int]:
+        """Sets unique seed for each environment.
+
+        Args:
+            seed (int): Seed number.
+
+        Raises:
+            AlreadyPendingCallError: If `seed` is called while another function call is pending.
+        """
+        self._assert_is_running()
+        seeds = [seed + i for i in range(self.num_envs)]
+
+        if self._state != AsyncState.DEFAULT:
+            raise AlreadyPendingCallError(
+                "Calling `seed` while waiting "
+                "for a pending call to `{0}` to complete.".format(self._state.value),
+                self._state.value,
+            )
+
+        for pipe, seed in zip(self.parent_pipes, seeds):
+            pipe.send(("seed", seed))
+        seeds, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+
+        return seeds
+
+    def reset_wait(
+        self, timeout: Union[int, float, None] = None
+    ) -> Sequence[Dict[str, Any]]:
+        """Waits for all environments to reset.
+
+        Args:
+            timeout (Union[int, float, None], optional): Seconds to wait before timing out.
+                Defaults to None, and never times out.
+
+        Raises:
+            NoAsyncCallError: If `reset_wait` is called without calling `reset_async`.
+            mp.TimeoutError: If response is not received from pipe within `timeout` seconds.
+
+        Returns:
+            Sequence[Dict[str, Any]]: A batch of observations from the vectorized environment.
+        """
+
+        self._assert_is_running()
+        if self._state != AsyncState.WAITING_RESET:
+            raise NoAsyncCallError(
+                "Calling `reset_wait` without any prior call to `reset_async`.",
+                AsyncState.WAITING_RESET.value,
+            )
+
+        if not self._poll(timeout):
+            self._state = AsyncState.DEFAULT
+            raise mp.TimeoutError(
+                "The call to `reset_wait` has timed out after "
+                "{0} second{1}.".format(timeout, "s" if timeout > 1 else "")
+            )
+
+        observations, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+        self._state = AsyncState.DEFAULT
+
+        return observations
```

Comment:
I see this, as well as other implementations here like `step_wait()`, are very similar to the implementations in the base class, but avoid using the `self.copy` and `self.shared_memory` members there (which would be public in this child class as welif the  `__init__` call did the normal thing and called `__init__` on the base class).  

What is the reason we're avoiding these optimizations?